Paintroid/src/main/java/org/catrobat/paintroid/command/implementation/CommandManagerImplementation.java

```diff
@@ -19,147 +19,513 @@
-import org.catrobat.paintroid.command.Command;
-import org.catrobat.paintroid.command.CommandManager;
-import org.catrobat.paintroid.command.UndoRedoManager;
-import org.catrobat.paintroid.command.UndoRedoManager.StatusMode;
-import org.catrobat.paintroid.dialog.IndeterminateProgressDialog;
+public class CommandManagerImplementation implements CommandManager, Observer
+{
+	private static final int INIT_APP_lAYER_COUNT = 1;
+
+	enum CommandType {COMMIT_LAYER_BITMAP_COMMAND
+		,ADD_LAYER
+		,REMOVE_LAYER
+		,MERGE_LAYERS
+		,CHANGE_LAYER_VISIBILITY
+		,LOCK_LAYER
+		,RENAME_LAYER}
+
 
-import android.graphics.Bitmap;
-import android.graphics.Bitmap.Config;
+	enum Action {UNDO, REDO}
 
-public class CommandManagerImplementation implements CommandManager, Observer {
-	private static final int MAX_COMMANDS = 512;
+	private LinkedList<Pair<CommandType, LayerCommand>> mLayerCommandList;
+	private LinkedList<Pair<CommandType, LayerCommand>> mLayerUndoCommandList;
+	private ArrayList<LayerBitmapCommand> mLayerBitmapCommands;
 
-	private final LinkedList<Command> mCommandList;
-	private int mCommandCounter;
-	private int mCommandIndex;
-	private Bitmap mOriginalBitmap;
+	private RefreshLayerDialogEventListener mRefreshLayerDialogListener;
+	private UpdateTopBarEventListener mUpdateTopBarListener;
+	private ArrayList<ChangeActiveLayerEventListener> mChangeActiveLayerListener;
+	private LayerEventListener mLayerEventListener;
+
+	public CommandManagerImplementation()
+	{
+		mLayerCommandList = new LinkedList<Pair<CommandType, LayerCommand>>();
+		mLayerUndoCommandList = new LinkedList<Pair<CommandType, LayerCommand>>();
+		mLayerBitmapCommands = new ArrayList<LayerBitmapCommand>();
+	}
 
-	public CommandManagerImplementation() {
-		mCommandList = new LinkedList<Command>();
-		// The first command in the list is needed to clear the image when
-		// rolling back commands.
-		mCommandList.add(new ClearCommand());
-		mCommandCounter = 1;
-		mCommandIndex = 1;
+	public void setRefreshLayerDialogListener(RefreshLayerDialogEventListener listener)
+	{
+		mRefreshLayerDialogListener = listener;
+	}
+
+	public void setUpdateTopBarListener(UpdateTopBarEventListener listener)
+	{
+		mUpdateTopBarListener = listener;
+	}
+
+	public void addChangeActiveLayerListener(ChangeActiveLayerEventListener listener)
+	{
+		if(mChangeActiveLayerListener == null)
+		{
+			mChangeActiveLayerListener = new ArrayList<ChangeActiveLayerEventListener>();
+		}
+
+		mChangeActiveLayerListener.add(listener);
+	}
+
+	public void setLayerEventListener(LayerEventListener listener)
+	{
+		mLayerEventListener = listener;
 	}
 
 	@Override
-	public boolean hasCommands() {
-		return mCommandCounter > 1;
+	public void commitCommandToLayer(LayerCommand layerCommand, Command bitmapCommand) {
+		synchronized (mLayerCommandList) {
+			clearUndoCommandList();
+			enableUndo(true);
+
+			ArrayList<LayerBitmapCommand> result = getLayerBitmapCommands(layerCommand.getLayer().getLayerID());
```

Comment:
why do you need a list with layerIDs and then commit the layer on the 0th? wouldn't it work to get the current layer id and just commit it there?